mmake.mk

```
@@ -444,9 +444,19 @@ $.project := $()
 $.project.targets = $(call $.get,$.project,targets)
 $.project.subprojects = $(call $.get,$.project,subprojects)
 
-# Shorthand for adding subprojects
-# (paths) -> (properties)
-$.add_subprojects = $(call $.set,subprojects,$1)
+# Adding subprojects
+# (paths)
+# In future (paths) -> (properties) ?
+define $.add_subprojects =
+$(eval $.mmake.make_all: $1)
+$(eval .PHONY: $1)
+$(call $.set,subprojects,$1)
+$(foreach p, $1,
+$(eval $p:
+>	$$(info Making subproject: $p)
+>	@$(MAKE) -C $(dir $p) -f $(notdir $p)
+))
```

Я бы такое форматирование предложил:
1) Лишние пробелы лучше не ставить даже если здесь они не важны. Кодовая база будет более однородной и снизится вероятность что случайно поставишь там где не надо.
2) Многострочный код для выполнения писать с новой строки, чтоб было нагляднее.
```suggestion
$(foreach p,$1,$(eval
$p:
>	$$(info Making subproject: $p)
>	$(MAKE) -C $(dir $p) -f $(notdir $p)
))
```
Ещё я здесь убрал `@`, зачем он?

---

@ нужно, чтобы не выводилась исполняемая команда. Если убрать, то будут сообщения типа `make -C ./ -f say_hi.mk`

---

А зачем не выводить команду? Она настолько сильно мешает? Мне всегда казалось что когда команды выводятся намного проще понимать что происходит

---

Не вижу смысла выводить команды. Если надо что-то выводить, то надо выводить прописанные логи, которые говорят что где происходит, исключительно релевантную информацию. Потому что потом искать где произошла ошибка с мусорными сообщениями команд тяжело. Я бы ещё убрал сообщения `make[1]: entering directory` если бы знал как.